### PR TITLE
Add support for timestamp to varchar coercer in hive tables

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description

Add support for timestamp to varchar coercer in hive tables. In case of partitioned table it is supported by most of the format and in case of unpartitioned tables only ORC format is supported as of now. 

The coercion that was supported as of current master for unpartitioned tables are inherently supported by the `ColumnReader` and this PR introduces a framework which maps a `OrcTypeKind` to a corresponding `TrinoType` and also re-uses the `TypeCoercer` used by partitioned tables. 

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Add support for timestamp to varchar coercer in hive tables
```

### CODE REVIEW

**Comment by Praveen2112** [comment]
Will be adding some additional test coverage for `TimestampCoercer` but wanted to have a early feedback here for the design/approach.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java [hunk 1]
```diff
@@ -43,37 +43,40 @@ private static HiveTableDefinition.HiveTableDefinitionBuilder tableDefinitionBui
...
-                            long_decimal_to_varchar            DECIMAL(20,12),
-                            short_decimal_to_bounded_varchar   DECIMAL(10,5),
-                            long_decimal_to_bounded_varchar    DECIMAL(20,12),
-                            varchar_to_bigger_varchar          VARCHAR(3),
-                            varchar_to_smaller_varchar         VARCHAR(3),
-                            char_to_bigger_char                CHAR(3),
-                            char_to_smaller_char               CHAR(3),
-                            id                                 BIGINT)
-                       STORED AS\s""" + fileFormat);
+                         CREATE TABLE %NAME%(
```
Comment (skrzypo987) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java:46]: This is looks like unintended format change. If that was the idea, please put it into a separate commit

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        public ShortTimestampToVarcharCoercer(TimestampType fromType, VarcharType toType)
+        {
+            super(fromType, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            long epochMicros = fromType.getLong(block, position);
```
Comment (skrzypo987) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:62]: Please add some unit test of that because of those `floorMods` and `floorDivs`. I've made mistake of thinking that those operations are super simple so no low-level testing is needed and, well, these are terrible to debug later on.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java [hunk 3]
```diff
@@ -326,17 +333,19 @@ public String toString()
             implements ColumnAdaptation
     {
         private final int index;
+        private final Function<Block, Block> blockCoercer;
 
-        public SourceColumn(int index)
+        public SourceColumn(int index, Function<Block, Block> blockCoercer)
```
Comment (skrzypo987) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java:338]: Function -> UnaryOperator

**Review by skrzypo987 — COMMENTED** [review]
Looks ok

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java (hunk 2 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:62]: I'm planning to expand the test coverage here but wanted to get some early feedback for the approach.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java (hunk 1 above)
Comment (Praveen2112) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java]: It was IntelliJ formatting messed this change. Thanks for pointing this one.

**Comment by Praveen2112** [comment]
@skrzypo987 AC

**Review by skrzypo987 — APPROVED** [review]
lgtm

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+public final class TimestampCoercer
+{
+    private static final DateTimeFormatter LOCAL_DATE_TIME = new DateTimeFormatterBuilder()
+            .parseCaseInsensitive()
+            .append(ISO_LOCAL_DATE)
+            .appendLiteral(' ')
+            .append(ISO_LOCAL_TIME)
+            .toFormatter()
+            .withResolverStyle(ResolverStyle.STRICT)
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:47]: is resolve used for formatting?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        public ShortTimestampToVarcharCoercer(TimestampType fromType, VarcharType toType)
+        {
+            super(fromType, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            long epochMicros = fromType.getLong(block, position);
+            long epochSecond = Math.floorDiv(epochMicros, MICROSECONDS_PER_SECOND);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:64]: import

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            super(fromType, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            LongTimestamp timestamp = (LongTimestamp) fromType.getObject(block, position);
+
+            long seconds = floorDiv(timestamp.getEpochMicros(), MICROSECONDS_PER_SECOND);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:88]: `epochSecond`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            long microsFraction = floorMod(timestamp.getEpochMicros(), MICROSECONDS_PER_SECOND);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:89]: inline (as above)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            long nanosFraction = (microsFraction * NANOSECONDS_PER_MICROSECOND) +
+                    (timestamp.getPicosOfMicro() / PICOSECONDS_PER_NANOSECOND);
+
+            toType.writeSlice(
+                    blockBuilder,
+                    truncateToLength(
+                            Slices.utf8Slice(
+                                    // Hive supports timestamp precision till nanoseconds to it is safe to cast to int.
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:97]: the comment is wrong. `nanosFraction` is at most 1e9

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                                    LOCAL_DATE_TIME.format(LocalDateTime.ofEpochSecond(seconds, (int) nanosFraction, UTC))),
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:98]: use `toIntExact` to avoid IDE warning

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            LongTimestamp timestamp = (LongTimestamp) fromType.getObject(block, position);
+
+            long seconds = floorDiv(timestamp.getEpochMicros(), MICROSECONDS_PER_SECOND);
+            long microsFraction = floorMod(timestamp.getEpochMicros(), MICROSECONDS_PER_SECOND);
+            long nanosFraction = (microsFraction * NANOSECONDS_PER_MICROSECOND) +
+                    (timestamp.getPicosOfMicro() / PICOSECONDS_PER_NANOSECOND);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:91]: add
```
// Hive timestamp has nanoseconds precision, so no truncation here

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        assertCoercions(TIMESTAMP_PICOS, new LongTimestamp(timestamp.getEpochMicros(), timestamp.getPicosOfMicros()), createVarcharType(27), "2023-04-11 05:16:12.3456788");
+        assertCoercions(TIMESTAMP_PICOS, new LongTimestamp(timestamp.getEpochMicros(), timestamp.getPicosOfMicros()), createVarcharType(28), "2023-04-11 05:16:12.34567887");
+        assertCoercions(TIMESTAMP_PICOS, new LongTimestamp(timestamp.getEpochMicros(), timestamp.getPicosOfMicros()), createVarcharType(29), "2023-04-11 05:16:12.345678876");
+    }
+
+    @DataProvider
+    public Object[][] timestampValuesProvider()
+    {
+        return new Object[][] {
+                //before epoch
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:97]:
```suggestion
                // before epoch
```

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java [hunk 12]
```diff
@@ -56,6 +57,7 @@ public void extendEnvironment(Environment.Builder builder)
                         "/docker/presto-init.d/presto-init-hdp3.sh");
             }
         });
+        builder.configureContainer(HADOOP, container -> container.withEnv("TZ", "/usr/share/zoneinfo/UTC"));
```
Comment (findepi) [inline testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java:60]: what is this for?

Comment (Praveen2112) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java:60]: When doing implicit coercion to timestamp, hive cast to a zone specific String - so wanted it to be in sync with UTC (like we do in Trino). This is only specific to ORC file for other formats it doesn't consider the timezone of the HiveServer.

Comment (findepi) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java:60]: quick google search suggests that `TZ` env variable means time zone name. `/usr/share/zoneinfo/UTC` looks like a file path, not a time zone name. How does it work?

> When doing implicit coercion to timestamp, hive cast to a zone specific String - so wanted it to be in sync with UTC (like we do in Trino).

would this information be useful for future code maintainers?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java (hunk 4 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:47]: I have copied this from ORC implementation - I think we can use it for timestamp -> varchar coercion

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java [hunk 13]
```diff
@@ -326,17 +333,19 @@ public String toString()
             implements ColumnAdaptation
     {
         private final int index;
+        private final UnaryOperator<Block> blockCoercer;
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java:336]: Should we mention any additional information as a description ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+import io.trino.plugin.hive.coercions.TypeCoercer;
+import io.trino.spi.type.TimestampType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.VarcharType;
+
+import java.util.Optional;
+
+import static io.trino.spi.type.TimestampType.createTimestampType;
+
+public class OrcTypeTranslator
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java:29]: Can you share the warning here - Mine is not showing any warning.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+import static io.trino.plugin.hive.coercions.CoercionAssertions.assertCoercions;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MICROS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_PICOS;
+import static io.trino.spi.type.VarcharType.createUnboundedVarcharType;
+import static io.trino.spi.type.VarcharType.createVarcharType;
+import static java.time.ZoneOffset.UTC;
+import static java.time.temporal.ChronoField.NANO_OF_SECOND;
+
+public class TestTimestampCoercer
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:31]: So for each coercion testing should we add explicit implementation ?

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java (hunk 12 above)
Comment (Praveen2112) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java:60]: Quick google search suggested me to set `TZ` to this file and it worked for me. Is there any workaround I could try here.

**Comment by Praveen2112** [comment]
@findepi AC

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        return Optional.empty();
+    }
+
+    public static Type fromOrcTypeKind(OrcTypeKind orcTypeKind, HiveTimestampPrecision timestampPrecision)
+    {
+        if (orcTypeKind.equals(OrcTypeKind.TIMESTAMP)) {
+            return createTimestampType(timestampPrecision.getPrecision());
+        }
+        throw new UnsupportedOperationException(orcTypeKind + " not yet supported");
+    }
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java]: I have added an API in `TypeCoercer` to provide us those details.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java (hunk 7 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:88]: Can you provide some context here.

**Comment by Praveen2112** [comment]
Will fix a few test failures

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java [hunk 17]
```diff
@@ -326,17 +333,19 @@ public String toString()
...
             checkArgument(index >= 0, "index is negative");
             this.index = index;
+            this.blockCoercer = requireNonNull(blockCoercer, "blockCoercer is null");
         }
 
         @Override
         public Block block(Page sourcePage, MaskDeletedRowsFunction maskDeletedRowsFunction, long filePosition, OptionalLong startRowId)
         {
-            Block block = sourcePage.getBlock(index);
+            Block block = blockCoercer.apply(sourcePage.getBlock(index));
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java:348]: I think this coercion should happen inside the `LazyBlockLoader#load` method on the result of `sourceBlock.getLoadedBlock()` to preserve laziness

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java:336]: It would be clearer to define a BlockTypeCoercion interface for `public Block apply(Block block)` and then use that in TypeCoercer instead of `Function<Block, Block>` and here instead of `UnaryOperator<Block>`.
It will also be easier to track possible implementations.

[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java:336]: The from and to types of the coercion would be useful info in the toString

**Comment by Praveen2112** [comment]
@findepi / @raunaqmorarka AC

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:42]: Implemented something similar to that.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java]: Instead of introducing another abstraction - We have created a new `ColumnAdaption` impl for coerced column.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TypeCoercer.java [hunk 18]
```diff
@@ -49,4 +49,14 @@ public Block apply(Block block)
     }
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TypeCoercer.java:49]: You can consider implementing an optimized path for dictionary block in TypeCoercer#apply. Also for RLE if that code can take a partitioned column's block as input.

Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TypeCoercer.java:49]: But `dictionary block` is applicable for varchar-varchar coercion right ?

Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TypeCoercer.java:49]: I was thinking on having it as a follow-up as it could benefit for other coercions too. WDYT ?

In case of `DictionaryBlock` is there is any specific reason on why `ids` is not exposed in a public API ?

Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TypeCoercer.java:49]: Doing it as follow-up is good.
You could add a public method to BlockUtil for applying transform to dictionary block and make use of non-public APIs internally in that if it makes sense.

**Comment by Praveen2112** [comment]
@raunaqmorarka AC (first one)

**Comment by Praveen2112** [comment]
@findepi / @raunaqmorarka Can you PTAL ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSourceProvider.java [hunk 19]
```diff
@@ -99,6 +100,7 @@ public HivePageSourceProvider(
         this.typeManager = requireNonNull(typeManager, "typeManager is null");
         this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
         this.domainCompactionThreshold = hiveConfig.getDomainCompactionThreshold();
+        this.timestampPrecision = hiveConfig.getTimestampPrecision();
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSourceProvider.java:103]: use `io.trino.plugin.hive.HiveSessionProperties#getTimestampPrecision`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+{
+    private static final DateTimeFormatter LOCAL_DATE_TIME = new DateTimeFormatterBuilder()
+            .parseCaseInsensitive()
+            .append(ISO_LOCAL_DATE)
+            .appendLiteral(' ')
+            .append(ISO_LOCAL_TIME)
+            .toFormatter()
+            .withChronology(IsoChronology.INSTANCE);
+
+    private TimestampCoercer() {}
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:49]: nit: move above all other members

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java (hunk 7 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:88]: above you have
```
int nanoFraction = floorMod(epochMicros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
```

without `microsFraction` intermediate variable

why not write the same logic using same style?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java [hunk 21]
```diff
@@ -364,9 +366,16 @@ else if (column.getBaseHiveColumnIndex() < fileColumns.size()) {
                 Type readType = column.getType();
                 if (orcColumn != null) {
                     int sourceIndex = fileReadColumns.size();
-                    columnAdaptations.add(ColumnAdaptation.sourceColumn(sourceIndex));
+                    Optional<TypeCoercer<?, ?>> coercer = createCoercer(orcColumn.getColumnType(), readType, getTimestampPrecision(session));
+                    if (coercer.isPresent()) {
+                        fileReadTypes.add(coercer.orElseThrow().getFromType());
+                        columnAdaptations.add(ColumnAdaptation.coercedColumn(sourceIndex, coercer.orElseThrow()));
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java:372]: nit: `orElseThrow` -> `get`

`orElseThrow` is recommended where you don't have `isPresent` check (for `get` there is an IDE warning)
`get` is shorter. yes, i know is has api-note "The preferred alternative to this method is `orElseThrow`". Well, Optionals are "recommended for method return type" and do we care?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+public final class OrcTypeTranslator
+{
+    private OrcTypeTranslator() {}
+
+    public static Optional<TypeCoercer<? extends Type, ? extends Type>> createCoercer(OrcTypeKind fromOrcType, Type toTrinoType, HiveTimestampPrecision timestampPrecision)
+    {
+        if (fromOrcType.equals(OrcTypeKind.TIMESTAMP) && toTrinoType instanceof VarcharType varcharType) {
+            TimestampType timestampType = (TimestampType) fromOrcTypeKind(fromOrcType, timestampPrecision);
+            if (timestampType.isShort()) {
+                return Optional.of(new ShortTimestampToVarcharCoercer(timestampType, varcharType));
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java:38]: Alternatively we could support to-varchar directly in `io.trino.orc.reader.TimestampColumnReader`,

no change requested (i like the current approach)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+            }
+            return Optional.of(new LongTimestampToVarcharCoercer(timestampType, varcharType));
+        }
+        return Optional.empty();
+    }
+
+    private static Type fromOrcTypeKind(OrcTypeKind orcTypeKind, HiveTimestampPrecision timestampPrecision)
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java:45]: inline, i.e. replace the call place with
```
TimestampType timestampType = createTimestampType(timestampPrecision.getPrecision());
```


(inlining was also requested in https://github.com/trinodb/trino/pull/16869#discussion_r1171027863 )

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java (hunk 12 above)
Comment (findepi) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java:61]: > Quick google search suggested me to set TZ to this file


interesting. i should have taken a note what was the search terms in my quick google search...

what happens if you
```
container.withEnv("TZ", "UTC")
```

(instead of the file)?

and what happens if you
```
container.withEnv("TZ", "/dev/null")
```

or
```
container.withEnv("TZ", "absolutely-random-string")
```
?

**Review by findepi — APPROVED** [review]

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java (hunk 12 above)
Comment (Praveen2112) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java]: It looks like specifying a timezone also works like passing the file
```
container.withEnv("TZ", "UTC")
```

If we pass 
```
container.withEnv("TZ", "/dev/null")
```

It switches to default value  `UTC` 

For 
```
container.withEnv("TZ", "absolutely-random-string")
```

It switches to default value `UTC`.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java (hunk 22 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java:38]: We could but `orc` is being used in iceberg connector also so we might be applying coercion (IIRC it is not supported by iceberg)

**Comment by Praveen2112** [comment]
@findepi AC

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java [hunk 24]
```diff
@@ -364,9 +366,16 @@ else if (column.getBaseHiveColumnIndex() < fileColumns.size()) {
                 Type readType = column.getType();
                 if (orcColumn != null) {
                     int sourceIndex = fileReadColumns.size();
-                    columnAdaptations.add(ColumnAdaptation.sourceColumn(sourceIndex));
+                    Optional<TypeCoercer<?, ?>> coercer = createCoercer(orcColumn.getColumnType(), readType, getTimestampPrecision(session));
+                    if (coercer.isPresent()) {
+                        fileReadTypes.add(coercer.orElseThrow().getFromType());
+                        columnAdaptations.add(ColumnAdaptation.coercedColumn(sourceIndex, coercer.get()));
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java:372]: would you mind using `get` in the `fileReadTypes.add(coercer.orElseThrow().getFromType());` line too?

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java (hunk 12 above)
Comment (findepi) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/configs/ConfigHdp3.java]: > If we pass
> 
> ```
> container.withEnv("TZ", "/dev/null")
> ```
> 
> It switches to default value `UTC`
> 
> For
> 
> ```
> container.withEnv("TZ", "absolutely-random-string")
> ```
> 
> It switches to default value `UTC`.

so `/usr/share/zoneinfo/UTC` looks like something meaningful, but actually is "garbage time zone" and causes something (the JVM?) to switch to UTC.

funny

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java (hunk 24 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java:372]: Oops missed that. Will change it

**Comment by Praveen2112** [comment]
Did a few minor changes on `TestTimestampCoercer` which could help us in https://github.com/trinodb/trino/pull/17071

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java [hunk 25]
```diff
@@ -392,14 +402,14 @@ public static boolean narrowerThan(CharType first, CharType second)
     {
         private final Optional<Function<Block, Block>> elementCoercer;
 
-        public ListCoercer(TypeManager typeManager, HiveType fromHiveType, HiveType toHiveType)
+        public ListCoercer(TypeManager typeManager, HiveType fromHiveType, HiveType toHiveType, HiveTimestampPrecision timestampPrecision)
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java:406]: Thanks for pointing it out.

**Comment by Praveen2112** [comment]
@huberty89 AC

**Comment by Praveen2112** [comment]
@raunaqmorarka Gentle ping

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+    public static void assertLongTimestampToVarcharCoercions(TimestampType fromType, LongTimestamp valueToBeCoerced, VarcharType toType, String expectedValue)
+    {
+        assertCoercions(fromType, valueToBeCoerced, toType, Slices.utf8Slice(expectedValue), NANOSECONDS);
+    }
+
+    public static void assertCoercions(Type fromType, Object valueToBeCoerced, Type toType, Object expectedValue, HiveTimestampPrecision timestampPrecision)
+    {
+        Block coercedValue = createCoercer(TESTING_TYPE_MANAGER, toHiveType(fromType), toHiveType(toType), timestampPrecision).orElseThrow()
+                .apply(nativeValueToBlock(fromType, valueToBeCoerced));
+        Assertions.assertThat(blockToNativeValue(toType, coercedValue))
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:146]: nit: static import `assertThat`

**Review by raunaqmorarka — APPROVED** [review]
lgtm
I'm wondering why this is specifically supported for ORC and not parquet or other file formats for unpartitioned columns. Does Apache Hive support this for non-orc formats ? If this coercion is to be supported regardless of file format, then we might want the coercion to be done in HivePageSource rather than OrcPageSource

**Comment by Praveen2112** [comment]
> why this is specifically supported for ORC and not parquet or other file formats for unpartitioned columns

Parquet doesn't support all types of coercion like ORC - For instance Parquet supports timestamp to varchar coercion but it doesn't support varchar to timestamp coercion, while both are supported by ORC formats. So to start with we are planning for orc and then we could extend it for other file formats. 


> then we might want the coercion to be done in HivePageSource rather than OrcPageSource

In that case we might need to introduce some sort of an abstraction that allows to Reader to provide the schema and compare with the `HivePageSource` in case of partitioned tables we compare the global schema with a local partition schema and we apply coercion policy. 

Implementing for ORC allows us to create some sort of a testing framework which could be utilized for other formats in the future.
